### DISCUSSION TITLE
openssl - alpine - zlib

### DISCUSSION THREAD

**Discussion by dgrollon1** [openssl/openssl#25837] [Q&A] [answered]
Hello there!!,

I need to use openssl with zlib, so by default it is not installed with zlib, i get src code and build it with zlib options. But when i execute it, return multiple errors of relocation.

Alpine package by default when execute command:

openssl cms -compress -in "fileIn.txt" -out "fileOut.txt"
Error creating CMS structure
288B85E7287F0000:error:17000097:CMS routines:CMS_compress:unsupported compression algorithm:crypto/cms/cms_smime.c:1001:

Decide to build it with zlib option, process execute:
1- Download src code and Start docker
docker run --rm -it -v /mnt/c/Users/user/Downloads/openssl-3.4.0.tar.gz:/opt/openssl-3.4.0.tar.gz alpine:3.20.3 sh
2- Install libraries
apk update
apk upgrade
apk add zlib zlib-dev perl musl-dev gcc make linux-headers
3- Decompress file
cd /opt/
tar -xzvf openssl-3.4.0.tar.gz
4- Compile and Build Openssl
cd openssl-3.4.0/
./Configure --prefix=/opt/openssl --openssldir=/usr/local/ssl --with-zlib-include=/lib/ zlib
make
make test
make install
5- Execute command
openssl cms -compress -in "fileIn.txt" -out "fileOut.txt"
6- Return output
Error relocating ./openssl: OSSL_CMP_ATAV_get0_algId: symbol not found
Error relocating ./openssl: OSSL_CMP_ATAV_get_rsaKeyLen: symbol not found
Error relocating ./openssl: OSSL_CMP_CRLSTATUS_get0: symbol not found
Error relocating ./openssl: PKCS12_set_pbmac1_pbkdf2: symbol not found
Error relocating ./openssl: OSSL_CMP_ITAV_get0_crlStatusList: symbol not found
Error relocating ./openssl: OSSL_CMP_ATAV_get0_type: symbol not found
Error relocating ./openssl: EVP_MD_xof: symbol not found
Error relocating ./openssl: TS_VERIFY_CTX_set0_imprint: symbol not found
Error relocating ./openssl: TS_VERIFY_CTX_set0_data: symbol not found
Error relocating ./openssl: OSSL_CMP_ATAVS_free: symbol not found
Error relocating ./openssl: TS_VERIFY_CTX_set0_certs: symbol not found
Error relocating ./openssl: OSSL_CMP_ITAV_new_crls: symbol not found
Error relocating ./openssl: SSL_get1_builtin_sigalgs: symbol not found
Error relocating ./openssl: TS_VERIFY_CTX_set0_store: symbol not found
Error relocating ./openssl: OSSL_CMP_get1_crlUpdate: symbol not found
Error relocating ./openssl: OSSL_CMP_ATAV_push1: symbol not found
Error relocating ./openssl: BIO_f_zlib: symbol not found
Error relocating ./openssl: OSSL_CMP_ITAV_new0_certReqTemplate: symbol not found
Error relocating ./openssl: OSSL_CMP_ATAV_new_rsaKeyLen: symbol not found
Error relocating ./openssl: OSSL_CMP_get1_certReqTemplate: symbol not found
Error relocating ./openssl: OSSL_CMP_ATAV_new_algId: symbol not found
Error relocating ./openssl: PBMAC1_get1_pbkdf2_param: symbol not found
Error relocating ./openssl: i2d_OSSL_CMP_ATAVS: symbol not found
Error relocating ./openssl: OSSL_CRMF_ATTRIBUTETYPEANDVALUE_free: symbol not found

**Comment by tmshort** [comment]
The problem appears to be that Alpine has its own install of OpenSSL, with the libraries in `/lib`, and this is what is being referenced by the `app/openssl` executable. The openssl libraries are being installed into `/opt/openssl/lib64`. Since `ldconfig` is not aware of `/opt/openssl/lib64`, it's dynamically loading the Alpine version of the libraries from `/lib`. Setting the `LD_LIBRARY_PATH=/opt/openssl/lib64` variable before running `apps/openssl` will make it work.
If you want a more permanent solution, you will want to set the installation prefix and libdir:

```
./Configure --openssldir=/usr/local/ssl --with-zlib-include=/lib/ zlib --prefix=/ --libdir=lib
```

This will overwrite Alpine's libraries. Otherwise, libdir will default to `lib64`. This will also place the `openssl` executable into  the `/bin` directory.

**Comment by mattcaswell** [comment]
> This will overwrite Alpine's libraries. 

Seems like a bad idea. Don't do that!

**Comment by tmshort** [comment, marked as answer]
> > This will overwrite Alpine's libraries.
> 
> Seems like a bad idea. Don't do that!

This is being done in a container, not the user's actual machine, so it may be what they want when building a container.

An alternative would be to use the `ldconfig` command to change the library search path, but as with @mattcaswell's warning, this would mean that your container would reference the new installation of OpenSSL, and ignore Alpine's libraries.

You should investigate the `-rpath` option, if you want to keep the Alpine OpenSSL and new OpenSSL separate:
```
./Configure --prefix=/opt/openssl --openssldir=/usr/local/ssl --with-zlib-include=/lib/ zlib -Wl,-rpath=/opt/openssl/lib64
make
make install_sw
```

Running  `apps/openssl` or `/opt/openssl/bin/openssl` will then work. You can confirm this via `ldd` of either of those executables.

Because there are now two sets of OpenSSL libraries, you need to explicitly specify which library to link against, that is what the `-rpath` option will do for you.

**Reply by dgrollon1** [reply]
Thanks!!, exactly we are doing it inside a container, and it only needs openssl to compress, sign and cipher as2 message. I will try this, thanks!!
